src/Builder/AbstractFormContractor.php

```diff
@@ -0,0 +1,195 @@
+                        .' Please create one for the target model: `%s`',
+                        $fieldDescription->getName(),
+                        $fieldDescription->getTargetModel()
+                    ));
+                }
+            }
+        } elseif ($this->isAnyInstanceOf($type, [AdminType::class])) {
+            if (!$fieldDescription->getAssociationAdmin()) {
+                throw new \RuntimeException(sprintf(
+                    'The current field `%s` is not linked to an admin.'
+                    .' Please create one for the target model : `%s`',
+                    $fieldDescription->getName(),
+                    $fieldDescription->getTargetModel()
+                ));
+            }
+
+            // NEXT_MAJOR: Change this check with: "if (!$fieldDescription->hasSingleValueAssociation())".
+            if (!$this->hasSingleValueAssociation($fieldDescription)) {
+                throw new \RuntimeException(sprintf(
+                    'You are trying to add `%s` field `%s` which is not One-To-One or  Many-To-One.'
+                    .' Maybe you want `%s` instead?',
+                    AdminType::class,
+                    $fieldDescription->getName(),
+                    CollectionType::class
+                ));
+            }
+
+            // set sensitive default value to have a component working fine out of the box
+            $options['btn_add'] = false;
+            $options['delete'] = false;
+
+            $options['data_class'] = $fieldDescription->getAssociationAdmin()->getClass();
+            $options['empty_data'] = static function () use ($fieldDescription) {
+                return $fieldDescription->getAssociationAdmin()->getNewInstance();
+            };
+            $fieldDescription->setOption('edit', $fieldDescription->getOption('edit', 'admin'));
+        // @phpstan-ignore-next-line
+        } elseif ($this->isAnyInstanceOf($type, [
+            CollectionType::class,
+            // NEXT_MAJOR: remove 'Sonata\CoreBundle\Form\Type\CollectionType'
+            'Sonata\CoreBundle\Form\Type\CollectionType',
+        ])) {
+            if (!$fieldDescription->getAssociationAdmin()) {
+                throw new \RuntimeException(sprintf(
+                    'The current field `%s` is not linked to an admin.'
+                    .' Please create one for the target model : `%s`',
+                    $fieldDescription->getName(),
+                    $fieldDescription->getTargetModel()
+                ));
+            }
+
+            $options['type'] = AdminType::class;
+            $options['modifiable'] = true;
+            $options['type_options'] = $this->getDefaultAdminTypeOptions($fieldDescription, $formOptions);
+        }
+
+        return $options;
+    }
+
+    abstract protected function hasAssociation(FieldDescriptionInterface $fieldDescription): bool;
+
+    abstract protected function hasSingleValueAssociation(FieldDescriptionInterface $fieldDescription): bool;
```

Comment:
If I understand correctly, this method will be removed in next major ? So a `NEXT_MAJOR` comment is needed.